plugins/ltac2/tac2entries.ml

```diff
@@ -629,18 +629,28 @@ type krule =
   (raw_tacexpr, _, 'act, Loc.t -> raw_tacexpr) Pcoq.Rule.t *
   ((Loc.t -> (Name.t * raw_tacexpr) list -> raw_tacexpr) -> 'act) -> krule
 
-let rec get_rule (tok : scope_rule token list) : krule = match tok with
+let keyword_needed dft_kw_need = function
+  | TacTerm _ :: _ -> false
+  | TacNonTerm _ :: _ -> true
+  | [] -> dft_kw_need
+
+let rec get_rule dft_kw_need (tok : scope_rule token list) : krule = match tok with
 | [] -> KRule (Pcoq.Rule.stop, fun k loc -> k loc [])
 | TacNonTerm (na, ScopeRule (scope, inj)) :: tok ->
-  let KRule (rule, act) = get_rule tok in
+  let KRule (rule, act) = get_rule true tok in
   let rule = Pcoq.Rule.next rule scope in
   let act k e = act (fun loc acc -> k loc ((na, inj e) :: acc)) in
   KRule (rule, act)
 | TacTerm t :: tok ->
-  let KRule (rule, act) = get_rule tok in
-  let rule = Pcoq.(Rule.next rule (Symbol.token (Pcoq.terminal t))) in
+  let kw_need = keyword_needed dft_kw_need tok in
+  let KRule (rule, act) = get_rule dft_kw_need tok in
   let act k _ = act k in
-  KRule (rule, act)
+  if kw_need then
```

Comment:
At the very least, the piece of code below should be factorized as
```ocaml
let tok =
  if kw_need then Pcoq.terminal t
  else match NumTok.Unsigned.parse_string t with
  | None -> Pcoq.terminal t
  | Some n -> Tok.PNUMBER (Some n)
in
KRule (Pcoq.(Rule.next rule (Symbol.token tok)), act)
```

Reply:
I now remember why I did not factorize. The problem is that the types are not the same in each branch, and we need `KRule` to encapsulate the different types in a common type.

Reply:
In that case, I think you should reduce the number of branches by merging the if-then-else and the match together, so that the two current calls to Pcoq.terminal are factorized into a single one.

Reply:
Yes, that'd be better. Done